Support synthetic fields to be linked by resteasy reactive links

Sgitario opened the pull request:
Synthetic fields are fields that are not part of the current entity to be linked, but still, we want these endpoints to be linked/exposed.

Before these changes, if we add a resource like the following one:

```java
@GET
@Path("/slugOrId/{slugOrId}")
@Produces(MediaType.APPLICATION_JSON)
@RestLink(rel = "get-by-slug-or-id")
public TestRecord getBySlugOrId(@PathParam("slugOrId") String slugOrId) {
    // ...
}
```

The path param `slugOrId` does not match with any field of the entity `TestRecord`, and because this resource is annotated with `RestLink`, it fails with a runtime exception saying that the field `slugOrId` was not found.

After these changes, we allow the resources that use non-existing fields as path params and still the link will be exposed as specified in the `RestLink` annotation.

geoand commented:
To be honest, I have no idea what this does :)

@FroMage since you know more about the links feature, care to comment?

Sgitario commented:
PR description updated with more details about the changes and the use case.

geoand commented:
Gotcha thanks.

I'll still delegate to @FroMage about this one :)

FroMage commented:
So, pffff. It appears this is not documented on https://quarkus.io/guides/rest-data-panache, I can't find any reference to `@RestLink` so I don't know what it did or did not do.
Documenting this in the guide would go a long way towards explaining what we're talking about here.
It sounds a bit like what RESTEasy did with `@LinkResource`, which is documented extensively at https://docs.jboss.org/resteasy/docs/6.2.1.Final/userguide/html/ch08.html#d5e791 but I'm not even sure.

geoand commented:
+100 on documentation

Sgitario commented:
Yep, documenting this extension is a must-have. I only found this limitation because I wanted to try something in one of my examples.

Sgitario commented:
I would not like to block these changes due to the lack of documentation on this extension, which should be done sooner than later in a separate pull request. 
What do you have in mind in regard to this pull request? @geoand @FroMage

geoand commented:
I'm fine with adding docs later, but I have 0️⃣ opinion on the actual functionality this brings, so I'll defer to @FroMage

FroMage commented:
Well, the problem is that without documentation I don't know what we're fixing here, so I can't have an informed opinion either :(

Sgitario commented:
At the moment, in Resteasy classic we can define links using unexisting fields. Example:

```java
@GET
    @Path("/get/{slugOrId}")
    @Produces({ MediaType.APPLICATION_JSON, "application/hal+json" })
    @LinkResource(rel = "get by slug or id", pathParameters = "slugOrId")
    public TestRecord getFirst(@PathParam("slugOrId") String slugOrId) {
        return RECORDS.get(0);
    }
```
This endpoint in Resteasy classic will produce the following response:

```json
{
                "id": 2,
                "slug": "second",
                "value": "Second value",
                "_links": {
                    "list": {
                        "href": "http://localhost:8081/records"
                    },
                    "get by slug or id": {
                        "href": "http://localhost:8081/records/get/slugOrId"
                    },
                    "first": {
                        "href": "http://localhost:8081/records/first"
                    }
                }
            }
```

Nevertheless, the same endpoint in Resteasy Reactive using the rr annotation `RestLink` will fail with a RuntimeException, which is what I'm trying to address here.

FroMage commented:
Huh, I guess this got added after I contributed that functionality. What is the purpose of returning `http://localhost:8081/records/get/slugOrId` to the client? That URI is invalid. I don't understand the point here.

Sgitario commented:
> Huh, I guess this got added after I contributed that functionality. What is the purpose of returning `http://localhost:8081/records/get/slugOrId` to the client? That URI is invalid. I don't understand the point here.

This URL is returned by resteasy classic. So, we can't change it here. However, I only added this example to state that this use case is supported in resteasy classic, but not in resteasy reactive, which is really the point. After these changes, for Resteasy reactive, we are returning the URL with the path param (.../get/{slugOrId}).

FroMage commented:
OK, so RESTEasy classic is not doing this right, and you're proposing to return a URI template as a link. Fine, I can imagine this being useful. To get back to the original code sample:

```java
@GET
@Path("/slugOrId/{slugOrId}")
@Produces(MediaType.APPLICATION_JSON)
@RestLink(rel = "get-by-slug-or-id", entityType = TestRecord.class)
public TestRecord getBySlugOrId(@PathParam("slugOrId") String slugOrId) {
    // ...
}
```

Why are we requiring `entityType = TestRecord.class`? This feels redundant with the method return type.

Sgitario commented:
> OK, so RESTEasy classic is not doing this right, and you're proposing to return a URI template as a link. Fine, I can imagine this being useful. To get back to the original code sample:
> 
> ```java
> @GET
> @Path("/slugOrId/{slugOrId}")
> @Produces(MediaType.APPLICATION_JSON)
> @RestLink(rel = "get-by-slug-or-id", entityType = TestRecord.class)
> public TestRecord getBySlugOrId(@PathParam("slugOrId") String slugOrId) {
>     // ...
> }
> ```
> 
> Why are we requiring `entityType = TestRecord.class`? This feels redundant with the method return type.

Yes, it's redudant and not necessary.

FroMage commented on extensions/resteasy-reactive/quarkus-resteasy-reactive-links/deployment/src/test/java/io/quarkus/resteasy/reactive/links/deployment/TestResource.java:
```diff
@@ -68,4 +68,16 @@ public TestRecord getBySlug(@PathParam("slug") String slug) {
+    @GET
+    @Path("/slugOrId/{slugOrId}")
+    @Produces(MediaType.APPLICATION_JSON)
+    @RestLink(rel = "get-by-slug-or-id", entityType = TestRecord.class)
```
Not sure if this should test setting the `entityType` or that it should work by default, given the return type.

Sgitario replied:
I've removed the `entityType` field as it was not necessary.

FroMage reviewed (changes requested):
LGTM, except this new addition should be documented in the user guide.

Sgitario commented:
@FroMage I've just updated the documentation of the Resteasy Reactive Links.

FroMage commented on docs/src/main/asciidoc/resteasy-reactive.adoc:
```diff
@@ -1354,35 +1354,27 @@ public class RecordsResource {
 }
 ----
-
-When calling the endpoint `/records` which is defined by the method `getAll` within the above resource using curl, you would get the web links header:
```
I don't think we should remove these sections, they're useful because they provide a global view of the results we generate, while the notes you added make them scattered. I mean, the notes and explanations are useful, but so is the global result. BTW, I myself would have included the JSON links as well, because I find link headers less useful.

Sgitario replied:
ok ok, I will revert it back and add a new section for these changes.

Sgitario replied:
updated

FroMage reviewed (changes requested):
Thanks for the new docs, but my opinion is that we should not remove the global result list of links. Now, if other reviewers disagree with me, I'll reconsider, but I really think it was useful.

Sgitario committed:
Support synthetic fields to be linked by resteasy reactive links

Synthetic fields are fields that are not part of the current entity to be linked, but still we want these endpoints to be linked/exposed.

See the working example for more details.

Sgitario commented:
@FroMage I've reverted the changes in the documentation and added a sentence explaining these changes. 
About using JSON for the web links, there is already a section for this: https://quarkus.io/guides/resteasy-reactive#json-hypertext-application-language-hal-support

FroMage reviewed (approved):
OK, great!
